### PR TITLE
DIT-758 Add auth filter

### PR DESCRIPTION
**Pull request by retroredge**

### CODE REVIEW

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import java.security.MessageDigest
+
+import com.google.common.io.BaseEncoding
+import org.scalatest.BeforeAndAfterEach
+import play.api.http.HttpVerbs
+import play.api.http.Status._
+import scalaj.http.Http
+import uk.gov.hmrc.bindingtarifffilestore.util.{BaseFeatureSpec, ResourceFiles}
+
+import scala.concurrent.duration.{FiniteDuration, _}
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: import scala.concurrent.duration._

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  private val timeout: FiniteDuration = 2.seconds
+  private val serviceUrl = s"http://localhost:$port"
+
+  feature("Authentication") {
+
+    scenario("Calling an endpoint with auth header present with correct value is allowed") {
+
+      When("I call an endpoint with the correct auth token")
+      val result =     Http(s"$serviceUrl/file")
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: extra spaces

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import com.google.common.io.BaseEncoding
+import org.scalatest.BeforeAndAfterEach
+import play.api.http.HttpVerbs
+import play.api.http.Status._
+import scalaj.http.Http
+import uk.gov.hmrc.bindingtarifffilestore.util.{BaseFeatureSpec, ResourceFiles}
+
+import scala.concurrent.duration.{FiniteDuration, _}
+
+class AuthSpec extends BaseFeatureSpec with ResourceFiles with BeforeAndAfterEach {
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: I think there is something wrong with the integration tests you wrote here
All tests assert that the response code is `403` - I believe we need to cover also happy scenarios

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    scenario("Incorrect hashed auth token value provided as query param to an endpoint is forbidden") {
+
+      When("I call the notify endpoint with a hash of auth token")
+      val result = Http(s"$serviceUrl/file?X-Api-Token=WRONG_VALUE")
+        .method(HttpVerbs.GET)
+        .asString
+
+      Then("The response code should not be forbidden")
```
Comment (gokyo) [inline test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: remove `not`

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala (hunk 3 above)
Comment (gokyo) [inline test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: if you do not mind, I would like to do some cleaning on this class

but if you do not agree, that's fine...

Comment (gokyo) [reply test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: I would simplify the sentences... they seem too complex now

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import javax.inject.{Inject, Singleton}
+import play.api.mvc.{Filter, RequestHeader, Result, Results}
+import uk.gov.hmrc.bindingtarifffilestore.config.AppConfig
+
+import scala.concurrent.Future
+
+@Singleton
+class AuthFilter @Inject()(appConfig: AppConfig)(implicit override val mat: Materializer) extends Filter {
+
+  override def apply(f: RequestHeader => Future[Result])(rh: RequestHeader): Future[Result] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: One thing we need to verify in this filter (and also the `binding-tariff-classification` filter) is to make sure that the `ping/ping` healthcheck endpoint is allowed by this filter.
If not, deployments to MDTP will fail....

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      BaseEncoding.base64Url().encode(MessageDigest.getInstance("SHA-256").digest(s.getBytes("UTF-8")))
+    }
+
+    val hashedTokenValue: String = hash(appConfig.authorization)
+    val headerValue: Option[String] = rh.headers.get("X-Api-Token")
+    val hashedQueryParamValues: String = rh.queryString.get("X-Api-Token").map(_.head).getOrElse("")
+
+    (headerValue, hashedQueryParamValues) match {
+      case (Some(appConfig.authorization), _) => f(rh)
+      case (None, `hashedTokenValue`) => f(rh)
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: In the tests you are not covering the scenario `case (Some(appConfig.authorization), `hashedTokenValue`)`
Your filter forbids it (I agree on it).
We need to add an extra test. I will do it for you

**Comment by gokyo** [comment]
I made some simplification on the tests
Let me know if you agree....

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala (hunk 5 above)
Comment (retroredge) [reply app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: Good spot. I've added code to the the filter to allow `/ping/ping` to be called without auth token

File: test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala (hunk 3 above)
Comment (retroredge) [reply test/it/uk/gov/hmrc/bindingtarifffilestore/AuthSpec.scala]: OK

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala (hunk 6 above)
Comment (retroredge) [reply app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: I TDD'd the filter so it seems unlikely that there is code in there that is not covered by the tests :)

Comment (retroredge) [reply app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: I've added a test for the opposite scenario to the extra one you've added and also added a test for the case when both header and query param are present. Talked these changes through with Ed.

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala (hunk 5 above)
Comment (retroredge) [reply app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: Done

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  private def hash: String => String = { s: String =>
+    BaseEncoding.base64Url().encode(
+      MessageDigest.getInstance("SHA-256").digest(s.getBytes("UTF-8"))
+    )
+  }
+
+  override def apply(f: RequestHeader => Future[Result])(rh: RequestHeader): Future[Result] = {
+
+    rh.uri match {
+      case uri if uri.contains(healthEndpointUri) => f(rh)
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: I would use a stricter check (exact match or at least `endsWith`)

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      case _ => ensureAuthTokenIsPresent(f, rh)
+    }
+  }
+
+  private def ensureAuthTokenIsPresent(f: RequestHeader => Future[Result], rh: RequestHeader) = {
+
+    val headerValue: Option[String] = rh.headers.get(authTokenName)
+    val hashedQueryParamValues: Option[String] = rh.queryString.get(authTokenName).map(_.head)
+
+    (headerValue, hashedQueryParamValues) match {
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: 👍

File: app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala (hunk 7 above)
Comment (retroredge) [reply app/uk/gov/hmrc/bindingtarifffilestore/filters/AuthFilter.scala]: Done

**Pull request merged**